core/src/main/web/app/controlpanel/controlpanelsessionitem-directive.js

```diff
@@ -25,6 +25,33 @@
 
   module.directive('bkControlPanelSessionItem', function(
       bkUtils, bkSession, bkCoreManager, bkRecentMenu, bkEvaluatePluginManager) {
+
+    function saveMostRecentNotebookContents(sessionId, pathInfo, format) {
+      var deferred = bkUtils.newDeferred();
+
+      $.cometd.subscribe('/latest-notebook-model', function(resp) {
+
+        var fileSaver = bkCoreManager.getFileSaver(pathInfo.uriType);
+        fileSaver.save(pathInfo.uri, resp.data.notebookJson)
+        .then(function() {
+          bkRecentMenu.recordRecentDocument(JSON.stringify({
+            uri: pathInfo.uri,
+            type: pathInfo.uriType,
+            readOnly: false,
+            format: _.isEmpty(format) ? '' : format
+          }));
+        })
+        .then(deferred.resolve)
+        .catch(deferred.reject);
+
+        $.cometd.unsubscribe('/latest-notebook-model');
+      });
+
+      $.cometd.publish('/request-latest-notebook-model', {sessionId: sessionId});
+
+      return deferred.promise;
+    }
+
```

Comment:
I approached this problem with the general assumption that the controlpanel has responsibility for saving the notebook. The reason for this approach was because the controlpanel has the responsibility over handing errors, issuing the final close command to all open notebook sessions, and providing feedback to the user about the operation.

If we were to change where the execution (save) was happening we would be essentially creating an RPC approach across windows / tabs.

Is this what we want?